### PR TITLE
TYP: Annotations

### PR DESCRIPTION
**Pull request by jbrockmendel**

### CODE REVIEW

File: pandas/core/indexes/base.py [hunk 1]
```diff
@@ -3569,6 +3569,7 @@ def _join_multi(self, other, how, return_indexers=True):
             # common levels, ldrop_names, rdrop_names
             dropped_names = ldrop_names + rdrop_names
 
+            assert isinstance(self, MultiIndex)  # for mypy
```
Comment (simonjayhawkins) [inline pandas/core/indexes/base.py:3572]: could move up to just below `if self_is_mi ...` or do away with `self_is_mi` and `other_is_mi`

File: pandas/core/arrays/base.py [hunk 2]
```diff
@@ -1167,6 +1167,10 @@ class ExtensionOpsMixin:
        with NumPy arrays.
     """
 
+    @classmethod
+    def _create_arithmetic_method(cls, op):
```
Comment (simonjayhawkins) [inline pandas/core/arrays/base.py:1171]: I recall that replacing
```
cls.__add__ = cls._create_arithmetic_method(operator.add)
```
etc. with
```
setattr(cls, "__add__",  cls._create_arithmetic_method(operator.add))
```

solves mypy issues, although I'm not getting failures if I remove this classmethod added here, so not sure what the mypy error is here.

which way do you prefer?

**Review by simonjayhawkins — COMMENTED** [review]
Thanks @jbrockmendel some initial comments (A typing PR is always good for a discussion :wink:)

File: pandas/core/algorithms.py [hunk 3]
```diff
@@ -1101,6 +1102,10 @@ def __init__(self, obj, n: int, keep: str):
         if self.keep not in ("first", "last", "all"):
             raise ValueError('keep must be either "first", "last" or "all"')
 
+    @abc.abstractmethod
+    def compute(self, method: str) -> FrameOrSeriesUnion:
```
Comment (simonjayhawkins) [inline pandas/core/algorithms.py:1106]: if you make SelectN generic and use `FrameOrSeries` typevar you can also get the return types for nlargest and nsmallest sorted. follow-on OK

File: pandas/core/internals/managers.py [hunk 4]
```diff
@@ -334,7 +334,7 @@ def reduce(self: T, func) -> T:
         # If 2D, we assume that we're operating column-wise
         assert self.ndim == 2
 
-        res_blocks = []
+        res_blocks: List[Block] = []
```
Comment (simonjayhawkins) [inline pandas/core/internals/managers.py:337]: mypy doesn't need this if refactored as 

```
res_blocks = [i for blk in self.blocks for i in blk.reduce(func)]
```

Revealed type is 'builtins.list[pandas.core.internals.blocks.Block*]

again, it's down to a personal preference of variable type annotations vs (in some cases complex) list comprehensions

Comment (jbrockmendel) [reply pandas/core/internals/managers.py:337]: I've got a couple other PRs that refactor this method, so not going to worry about this too much

File: pandas/core/arrays/base.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/core/arrays/base.py:1171]: i prefer the aesthetics of the `cls.__add__ =` variant, but not a strong enough preference to make a fuss over

File: pandas/core/internals/managers.py (hunk 4 above)
Comment (simonjayhawkins) [reply pandas/core/internals/managers.py:337]: sure

File: pandas/core/arrays/base.py (hunk 2 above)
Comment (simonjayhawkins) [reply pandas/core/arrays/base.py:1171]: sure

**Review by simonjayhawkins — APPROVED** [review]
Thanks @jbrockmendel

**Comment by simonjayhawkins** [comment]
Thanks @jbrockmendel

**Pull request merged**